### PR TITLE
Allow calling standard completion function from custom one

### PR DESCRIPTION
**Pull request by gardenappl**
See #1013 

Based on work by @tkarls (https://github.com/yargs/yargs/issues/1013#issuecomment-346814408). This is my first time working with TypeScript so my solution might be a bit ugly.

One thing that worries me is that this API requires you to use callbacks. There is no good way to overload this function:

```ts
export type CompletionFunction =
  | SyncCompletionFunction
  | AsyncCompletionFunction
  | FallbackCompletionFunction;

interface SyncCompletionFunction {
  (current: string, argv: Arguments): string[] | Promise<string[]>;
}

interface AsyncCompletionFunction {
  (current: string, argv: Arguments, done: (completions: string[]) => any): any;
}

interface FallbackCompletionFunction {
  (
    current: string,
    argv: Arguments,
    defaultCompletion: () => any,
    done: (completions: string[]) => any
  ): any;
}
```

Each function takes 2, 3 or 4 arguments respectively.

One possible solution: since you're about to release a new major version, it might be OK to just remove the ability to use callbacks, in favor of Promises. So then the new API would look like this:

```ts
export type CompletionFunction =
  | LegacyCompletionFunction
  | FallbackCompletionFunction;

interface LegacyCompletionFunction {
  (current: string, argv: Arguments): string[] | Promise<string[]>;
}

interface FallbackCompletionFunction {
  (current: string, argv: Arguments, defaultCompletion: () => any): string[] | Promise<string[]> | void;
}
```

### CODE REVIEW

File: lib/completion.ts [hunk 1]
```diff
@@ -34,6 +34,78 @@ export function completion(
     const current = args.length ? args[args.length - 1] : '';
     const argv = yargs.parse(args, true);
     const parentCommands = yargs.getContext().commands;
+
+    function defaultCompletion(): Arguments | void {
+      const handlers = command.getCommandHandlers();
```
Comment (bcoe) [inline lib/completion.ts:39]: i would be tempted to pull each of these major pieces of logic into a helper function within `completion.ts`:

* `commandCompletions`?
* `optionCompletions`?
* `customCompletions`?

File: lib/completion.ts [hunk 2]
```diff
@@ -34,6 +34,78 @@ export function completion(
...
+      if (current.match(/^-/) || (current === '' && completions.length === 0)) {
+        const descs = usage.getDescriptions();
+        const options = yargs.getOptions();
+        Object.keys(options.key).forEach(key => {
+          const negable =
+            !!options.configuration['boolean-negation'] &&
+            options.boolean.includes(key);
+          // If the key and its aliases aren't in 'args', add the key to 'completions'
+          let keyAndAliases = [key].concat(aliases[key] || []);
+          if (negable)
```
Comment (bcoe) [inline lib/completion.ts:75]: It's this block especially where I think we'd benefit from a bit more decomposition. Could `completeOptionKey` could be pulled out to the top level of the file, along with us breaking apart `defaultCompletion` into a couple more helpers?

I think if we then add comments to each of the helpers, the completion logic might become easier to consume.

File: lib/completion.ts [hunk 3]
```diff
@@ -190,8 +207,23 @@ interface AsyncCompletionFunction {
...
+  ): any;
+}
+
 function isSyncCompletionFunction(
   completionFunction: CompletionFunction
 ): completionFunction is SyncCompletionFunction {
   return completionFunction.length < 3;
 }
+
+function isFallbackCompletionFunction(
```
Comment (bcoe) [inline lib/completion.ts:225]: I like the use of type detection here 👍

File: lib/completion.ts [hunk 4]
```diff
@@ -190,8 +207,23 @@ interface AsyncCompletionFunction {
...
+  completionFunction: CompletionFunction
+): completionFunction is FallbackCompletionFunction {
+  return completionFunction.length > 3;
```
Comment (bcoe) [inline lib/completion.ts:228]: should this be `completionFunction.arguments.length` I'm surprised this worked.

File: test/completion.cjs [hunk 5]
```diff
@@ -340,6 +340,48 @@ describe('Completion', () => {
...
+          if (err) throw err;
+          r.logs.should.include('apple');
+          r.logs.should.include('foo');
+          return done();
+        }
+      );
+    });
+
+    it('allows calling callback instead of default completion function', done => {
+      checkUsage(
```
Comment (bcoe) [inline test/completion.cjs:365]: tests read quite well 👍

**Review by bcoe — CHANGES_REQUESTED** [review]
@gardenappl this is looking really solid to me. What jumps out at me the most is that we've collected enough logic in `completion.ts`, that I think we'd benefit from a tiny bit of refactoring.

**Comment by gardenappl** [comment]
I tried splitting the code into multiple functions, but they required a lot of arguments and it didn't look good at all. So instead I spent some time refactoring the whole thing into a class.
I also fixed handling of short options in aliases (see changes in unit tests).

**Comment by gardenappl** [comment]
I am still kinda bothered by the fact that the API itself is not very clean. Right now the behaviour of `completion(...)` depends on how many arguments the function accepts:
* 2 args = synchronous function or Promise function, can't call default
* 3 args = callback function, can't call default
* 4 args = callback function, can call default

But this means that if you want to call the default function then you have to use callbacks. I already mentioned this when I created the PR and would like to hear some feedback on that.

**Review by bcoe — APPROVED** [review]
I think pulling the logic into a class was a very good compromise... and approach we could perhaps use elsewhere in the codebase.

LGTM. I will merge and release to a `next` branch so that you can test.

**Pull request merged**